File: src/main/java/com/wesker926/leetcode/algorithms/p0260/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0260.s1;

/**
 * @author wesker.gh
 * @date 2021/10/30
 * @description xor & lowbit
 */
public class Solution {
    public int[] singleNumber(int[] nums) {
        int xor = 0;
        for (int num : nums) xor ^= num;
        int lowbit = xor & (-xor);
        int[] result = new int[2];
        for (int num : nums) result[(num & lowbit) != 0 ? 0 : 1] ^= num;
        return result;
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0260/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/10/30
// @description: xor
package s1

func singleNumber(nums []int) []int {
	xor, bit := 0, 0
	for _, n := range nums {
		xor ^= n
	}
	bit = (xor & (xor - 1)) ^ xor

	result := make([]int, 2)
	for _, n := range nums {
		if (n & bit) != 0 {
			result[0] ^= n
		} else {
			result[1] ^= n
		}
	}
	return result
}
